### DISCUSSION TITLE
JWT와 인증 인가

### DISCUSSION THREAD

**Discussion by greenblues1190** [woowacourse-study/2022-woowahan-bansanghwe#123] [Network]
## JWT

> JSON Web Token (JWT) is a compact, URL-safe means of representing claims to be transferred between two parties. The claims in a JWT are encoded as a JSON object that is used as the payload of a JSON Web Signature (JWS) structure or as the plaintext of a JSON Web Encryption (JWE) structure, enabling the claims to be digitally signed or integrity protected with a Message Authentication Code (MAC) and/or encrypted. - [*RFC 7519*](https://datatracker.ietf.org/doc/html/rfc7519)

토큰 기반 인증. private key를 이용하여 서명 혹은 암호화한 json 객체 형태의 토큰을 클라이언트에 저장하여 인증한다.

### JWT의 구조

JWT은 `.`으로 구분되는 세 부분으로 이루어지며 Base64Url로 인코딩된다. `xxxxx.yyyyy.zzzzz`

#### Header

토큰의 타입(JWT)과 사용된 서명 알고리즘(HMAC SHA256 or RSA)이 담겨있다. 

```json
{
  "alg": "HS256",
  "typ": "JWT"
}
```

#### Payload

Claim(주장이라고 번역하는 것 같다)들이 담겨있다. Claim은 사용자와 같은 엔티티와 추가적인 정보에 대한 진술이다. Claim에는 세가지 종류가 있다.

- [Registered claims](https://tools.ietf.org/html/rfc7519#section-4.1): 필수는 아니지만 권장되는 미리 정의된 Claims. **iss** (issuer), **exp** (expiration time), **sub** (subject), **aud** (audience), [등등](https://tools.ietf.org/html/rfc7519#section-4.1)이 있다.
  > Notice that the claim names are only three characters long as JWT is meant to be compact.
- [Public claims](https://tools.ietf.org/html/rfc7519#section-4.2): JWT를 사용하는 사람들이 마음대로 정의할 수 있다. 그러나 충돌을 방지하려면[ IANA JWT 레지스트리](https://www.iana.org/assignments/jwt/jwt.xhtml)에 정의하거나 충돌 방지 네임스페이스를 포함하는 URI로 정의해야한다.
- [Private claims](https://tools.ietf.org/html/rfc7519#section-4.3): Registered claims이나 public claims가 아닌 당사자 간에 정보를 공유하기 위해 생성된 커스텀 Claim이다.

payload의 예는 다음과 같다.

```
{
  "sub": "1234567890",
  "name": "John Doe",
  "admin": true
}
```

#### Signature

signature 부분을 생성하려면 인코딩된 header, 인코딩된 payload, 암호, header에 지정된 알고리즘으로 서명해야 한다. 예를 들어 HMAC SHA256 알고리즘을 사용하려는 경우 서명은 다음과 같은 방식으로 생성된다.

```
HMACSHA256(
  base64UrlEncode(header) + "." +
  base64UrlEncode(payload),
  secret)
```

서명은 메시지가 도중에 변경되지 않았는지 확인하는 데 사용되며 개인 키로 서명된 토큰의 경우 JWT의 보낸 사람이 누구인지 확인할 수 있다.

### JWT의 동작 방식

![JSON 웹 토큰의 작동 방식](https://images.ctfassets.net/kbkgmx9upatd/2qcBkSHR7Rfs0UIZYhjd7k/b428b00508dca46d67f48c81019cdcc8/jwt-diagram.png)

- 사용자가 성공적으로 로그인한다.
- 서버는 사용자에게 토큰을 반환한다.
- 사용자가 보호된 경로 또는 자원에 접근하려고 할 때마다 사용자 에이전트는 일반적으로 Bearer 스키마 를 사용하여 Authorization 헤더 에서 JWT를 보낸다. `Authorization: Bearer <token>`

#### 주의할 점

- JWT를 필요 이상으로 오래 보관하지 않는다.
- [보안이 취약하기 때문에 민감한 세션 데이터를 브라우저 저장소에 저장하지 않는다.](https://cheatsheetseries.owasp.org/cheatsheets/HTML5_Security_Cheat_Sheet.html#local-storage)
- HTTP 헤더를 통해 JWT 토큰을 보내는 경우 토큰이 너무 커지지 않도록 한다. 일부 서버는 8KB 이상의 헤더를 허용하지 않는다.

## refresh token

https://auth0.com/docs/secure/tokens/refresh-tokens

## 참고 문서

https://jwt.io/introduction/

**Comment by kamwoo** [comment]
자기소개를 해주세요

**Comment by al-bur** [comment]
저희 회사에 지원하려고 하기위해 어떠한 노력을 하고 있나요?

**Comment by kamwoo** [comment]
인증/인가에 대해서 아시는대로 설명해주세요

**Comment by kamwoo** [comment]
프로젝트에서 인증/인가를 구현한 경험

**Comment by al-bur** [comment]
보안이 중요한 곳에서 인증/인가를 구현한 경험

**Comment by al-bur** [comment]
JWT?

**Reply by al-bur** [reply]
인증/인가때만 사용되나요?

**Comment by kamwoo** [comment]
refresh 토큰은 localStorage에 저장한다면 보안상의 문제점은 없을까요?

**Comment by kamwoo** [comment]
https를 사용한다면 http body에 아이디, 비밀번호를 실어서 보내도 될까요?

**Comment by al-bur** [comment]
refresh token을 도입한 이유는?

**Reply by al-bur** [reply]
5분마다 accessToken을 재발급하는 것에 대한 트레이드 오프는?

**Reply by al-bur** [reply]
5분이라는 시간에 대한 객관적인 지표는?

**Comment by kamwoo** [comment]
localStorage와 Cookie의 보안상의 차이점
